File: lib/src/models/product.dart
import 'package:sneakerx/src/models/product_image.dart';

class Product {
  final int productId;
  final int categoryId;
  final int shopId;
  final String name;
  final String description;
  final double basePrice;
  final double rating;
  final int soldCount;
  final DateTime createdAt;

  final List<ProductImage> images;

  Product({
    required this.productId,
    required this.shopId,
    required this.categoryId,
    required this.name,
    required this.description,
    required this.basePrice,
    required this.rating,
    required this.soldCount,
    required this.images,
    required this.createdAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      productId: json['productId'],
      shopId: json['shopId'],
      categoryId: json['categoryId'],
      name: json['name'],
      description: json['description'],
      basePrice: json['basePrice'],
      rating: json['rating'].toDouble(),
      soldCount: json['soldCount'],
      createdAt: DateTime.parse(json['createdAt']),
      images: (json['images'] as List?)
        ?.map((image) => ProductImage.fromJson(image as Map<String, dynamic>))
        .toList() ?? []
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'shopId': shopId,
      'categoryId': categoryId,
      'name': name,
      'description': description,
      'rating': rating,
      'status': "ACTIVE",
      'soldCount': soldCount,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: src/main/java/com/example/sneakerx/entities/Product.java
package com.example.sneakerx.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "products")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer productId;

    @ManyToOne
    @JoinColumn(name = "shop_id")
    @JsonIgnore // <--- THE FIX: Stops Jackson from serializing the Cart back to the parent
    @ToString.Exclude // <--- Stop Lombok from printing it
    @EqualsAndHashCode.Exclude
    private Shop shop;

    @ManyToOne
    @JoinColumn(name = "category_id")
    @JsonIgnore // <--- THE FIX: Stops Jackson from serializing the Cart back to the parent
    @ToString.Exclude // <--- Stop Lombok from printing it
    @EqualsAndHashCode.Exclude
    private Category category;

    private String name;
    private String description;
    private Double basePrice;
    private Integer soldCount;
    private Double rating;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductImage> images;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductSku> skus;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductReview> reviews;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductAttribute> attributes;
}
